test/helpers/position_assertions.cc

```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
+    if (regex_search(assertionContentsString, matches, newNameRegex)) {
+        auto version = matches[1].str();
+        auto newName = matches[2].str();
+        auto invalid = !matches[3].str().empty();
+        auto expectedErrorMessage = matches[4].str();
+        if (!newName.empty() || invalid) {
+            return make_shared<ApplyRenameAssertion>(filename, range, assertionLine, version, newName, invalid,
+                                                     expectedErrorMessage);
+        }
+    }
+
+    ADD_FAIL_CHECK_AT(string(filename).c_str(), assertionLine + 1,
+                      fmt::format("Improperly formatted apply-rename assertion. Expected '[<version>] newName: <name> "
+                                  "(invalid: true)'. Found '{}' in file {}",
+                                  assertionContents, filename));
+
+    return nullptr;
+}
+
+ApplyRenameAssertion::ApplyRenameAssertion(string_view filename, unique_ptr<Range> &range, int assertionLine,
+                                           string_view version, string newName, bool invalid,
+                                           string expectedErrorMessage)
+    : RangeAssertion(filename, range, assertionLine), version(string(version)), newName(newName), invalid(invalid),
+      expectedErrorMessage(expectedErrorMessage) {}
+
+void ApplyRenameAssertion::checkAll(const vector<shared_ptr<RangeAssertion>> &assertions,
+                                    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                    LSPWrapper &wrapper, int &nextId, string errorPrefix) {
+    for (auto assertion : assertions) {
+        if (auto assertionOfType = dynamic_pointer_cast<ApplyRenameAssertion>(assertion)) {
+            assertionOfType->check(sourceFileContents, wrapper, nextId, errorPrefix);
+        }
+    }
+}
+
+void ApplyRenameAssertion::check(const UnorderedMap<std::string, std::shared_ptr<core::File>> &sourceFileContents,
+                                 LSPWrapper &wrapper, int &nextId, std::string errorPrefix) {
+    auto prepareRenameResponse = doTextDocumentPrepareRename(wrapper, *this->range, nextId, this->filename);
+    // TODO: test placeholder in prepareRenameResponse
+
+    // A rename at an invalid position
+    if (newName.empty() && invalid) {
+        REQUIRE_EQ(prepareRenameResponse, nullptr);
+    } else {
+        REQUIRE_NE(prepareRenameResponse, nullptr);
+    }
+
+    auto workspaceEdits =
+        doTextDocumentRename(wrapper, *this->range, nextId, this->filename, newName, expectedErrorMessage);
+    // A rename at a valid position but with an invalid new name
+    if (invalid) {
+        REQUIRE_EQ(workspaceEdits, nullptr);
+        return;
+    }
+
+    {
+        INFO("doTextDocumentRename failed; see error above.");
+        REQUIRE_NE(workspaceEdits, nullptr);
+    }
+
+    auto &renameItems = *workspaceEdits->documentChanges;
+    auto it = sourceFileContents.find(this->filename);
```

Comment:
Minor nit: Is this variable used anywhere? I think you can replace this with an `if (sourceFileContents.find(this->filename) == sourceFileContents.end() {`